### ISSUE TITLE
HAL-06 iteration over map source of non-determinism

### ISSUE THREAD

**Issue by juniuszhou** [Sifchain/sifnode#3257]
# Description
The codebase contains instances of iteration over maps. As maps are not ordered in Go, iterations over maps are non-deterministic. If the consensus logic expects a deterministic result from the iteration, errors can occur where different nodes reach different states due to the differences in map ordering. If the nodes reach differing states, this can cause consensus issues, which could result in a chain halt.

Code Location:
```
x/tokenregistry/types/types.pb.go, Lines 582-601
Listing 8: Iteration over map (Line 582)
582 for k := range m. DoublePeggedNetworkMap {
583 v := m. DoublePeggedNetworkMap [k]
584 baseI := i
585 i- -
586 if v {
587 dAtA [i] = 1
588 } else {
589 dAtA [i] = 0
590 }
591 i- -
592 dAtA [i] = 0 x10
593 i = encodeVarintTypes (dAtA , i, uint64 (k))
594 i- -
595 dAtA [i] = 0x8
596 i = encodeVarintTypes (dAtA , i, uint64 (baseI -i))
597 i- -
598 dAtA [i] = 0x1
599 i- -
600 dAtA [i] = 0 x9a
601 }
```

# Recommendation
Ensure that iterations over maps do not contain any logic that depends on
a fixed ordering of the map. Alternatively, make use of an alternative
data structure that has a deterministic ordering or ensure that maps are
sorted before iterating over them.

**Comment by juniuszhou** [comment]
I don't agree with the report for the issue. Map is not determistic, but if use the smae map realization and the same order of updates, then the result should be same. If we use the vector as suggested, it will be no efficient for date fetch. I suggest we won't fix it.

**Comment by juniuszhou** [comment]
close it if one more person agree with no fix. @banshee @smartyalgo @Brando753

**Comment by juniuszhou** [comment]
After discuss with @smartyalgo , we think it is an issue, we should avoid using the map in keeper. will fix it.

**Comment by pandaring2you** [comment]
unit + integration tests modified and finished. PR awaiting review.

**Comment by pandaring2you** [comment]
@Brando753 to review PR

**Comment by pandaring2you** [comment]
@banshee reviewing. ETA next week.

**Comment by pandaring2you** [comment]
According to Michael: James is reorganizing solution to use keepers instead of maps.

**Comment by pandaring2you** [comment]
@juniuszhou to check with @banshee to see if he's currently working on a PR

**Comment by pandaring2you** [comment]
@juniuszhou to check in with James
